ic-identity-hsm/src/hsm.rs

```diff
@@ -129,16 +133,20 @@ fn open_session(
 }
 
 // We might need to log in.  This requires the PIN.
-fn login_if_required(
+fn login_if_required<PinFn>(
     ctx: &Ctx,
     session_handle: CK_SESSION_HANDLE,
-    pin: &str,
+    pin_fn: PinFn,
     slot_id: CK_SLOT_ID,
-) -> Result<bool, HardwareIdentityError> {
+) -> Result<bool, HardwareIdentityError>
+where
+    PinFn: FnOnce() -> Result<String, String>,
```

Comment:
Just to make sure, this is only going to be called once, right?

Reply:
That's right.  We don't even hold onto a reference to the function.